rosbag2_transport/src/rosbag2_transport/player_service_client.cpp

```diff
@@ -225,10 +225,28 @@ bool PlayerServiceClientManager::register_request_future(
 }
 
 bool PlayerServiceClientManager::wait_for_all_futures(std::chrono::duration<double> timeout)
+{
+  return wait_for_one_client_all_futures(nullptr, timeout);
+}
+
+bool PlayerServiceClientManager::wait_for_one_client_all_futures(
+  std::shared_ptr<rclcpp::GenericClient> specified_client,
+  std::chrono::duration<double> timeout)
 {
   auto is_all_futures_ready = [&]() {
       bool is_ready = true;
       for (auto & [timestamp, future_request_id_and_client] : request_futures_list_) {
+        // If service client is specified, find it
+        if (specified_client != nullptr) {
+          auto client = future_request_id_and_client.second.lock();
+          if (client == nullptr) {
+            throw std::runtime_error("request's client is not valid\n");
+          }
+          if (specified_client != client) {
+            continue;
+          }
+        }
+
```

Comment:
@Barry-Xu-2018 I don't get it. 
Why it is a matter to make sure that client is alive here?
We actually shouldn;t be care about client in the scope of this function only about future.

Reply:
Honestly, I would revert this last commit. 
I don't see any benifits from it only overcomplicated logic and extra API which is confusing.

Otherwise please explain.

Reply:
@Barry-Xu-2018 Never mind I have figured out. I will clean up API and usage of those functions.

Reply:
@MichaelOrlov i was going to post the note, but thanks!